remote/remote_test.go

```diff
@@ -168,3 +171,84 @@ func doTestWatch(t *testing.T, sm subnet.Manager) {
 		t.Errorf("WatchSubnet produced wrong subnet: expected %s, got %s", l.Key(), evt.Lease.Key())
 	}
 }
+
+func doTestWatchNetworks(t *testing.T, sm subnet.Manager, serverRegistry *subnet.MockSubnetRegistry) {
+	ctx, cancel := context.WithCancel(context.Background())
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	defer func() {
+		cancel()
+		wg.Wait()
+	}()
+
+	// Because remote clients are not allowed to add/remove networks, there
+	// is no server API for these operations.  But testing network watches
+	// still requires this, and that must happen directly in the server's
+	// registry.  That can cause race conditions if the WatchNetworks
+	// goroutine hasn't been run yet by the time CreateNetwork is called,
+	// so testing operations that touch the server registry must be deferred
+	// a short time to ensure that the server is listening when they happen.
+
+	events := make(chan []subnet.Event)
+	go func() {
+		subnet.WatchNetworks(ctx, sm, events)
+		wg.Done()
+	}()
+
+	// skip over the initial snapshot
+	<-events
+
+	expectedNetname := "foobar"
+	go func() {
+		// Add a new network a short time later to ensure that the
+		// client-side network watch triggers a server watch before the
+		// network is added directly in the server with CreateNetwork()
+		time.Sleep(time.Second)
```

Comment:
Your approach of directly modifying the server's register is just fine -- you're just simulating an external event.

> The server needs to be watching before the new network is created on the client side, otherwise the server-side will completely miss the event.

This is the scary statement. `subnet.WatchNetworks` internally maintains a cursor so it should not miss an event. If `CreateNetwork` executes before `etcd.go::WatchNetworks()`, that should be fine. When it does get to `etcd.go::WatchNetworks()`, the cursor will be pointing slightly behind the "end" and it will immediately return the added network. That is the whole point of maintaining and propagating the cursor.